### PR TITLE
Make `@Stepwise` work for iterated features

### LINKED ISSUE #1008: Allow spec iterations to be skipped
# Feature request

Spock currently does not currently support natively skipping a single iteration of an examples-based feature. The workaround mentioned in the documentation involves using JUnit's `AssumptionViolatedException` to handle such cases.

Spock already correctly [reports such iterations as "skipped" to the JUnit `Notifier`](https://github.com/spockframework/spock/blob/master/spock-core/src/main/java/org/spockframework/runtime/JUnitSupervisor.java#L92), but not to Spock's own `IRunListener` (which, incidentally, causes a [bug in spock-reports](https://github.com/renatoathaydes/spock-reports/issues/89) as it cannot know that a particular iteration of a feature was skipped).

It would be ideal to be able to use a Spock, backend-test-framework-independent way of doing this. Specifically, Spock's `IRunListener` should have a method, `void iterationSkipped( IterationInfo iteration )`, analogous to the existing `void specSkipped( SpecInfo spec )`.

## Proposed solution 1

I would suggest adding a method to `Specification` like `void skipIf(boolean)` and/or `assumeThat(boolean)`, that causes the current feature/iteration to be skipped when necessary.

## Proposed solution 2

It might be more Spock-like to use a `precondition` block instead:

```groovy
precondition: 'must be on Windows'
OS.isWindows()
```

# Example

Here's a minimal specification demonstrating how feature iterations can be skipped today, but only when using JUnit APIs.

```groovy
import org.junit.Assume
import spock.lang.Specification
import spock.lang.Unroll

class AssumptionViolatedTest extends Specification {
    @Unroll
    def "Can ignore test if AssumptionViolationException is thrown"() {
        setup: 'check some precondition'
        Assume.assumeFalse( i == 2 )
        expect:
        i > 0
        where:
        i << [ 1, 2, 3 ]
    }
}
```

With the first proposed solution, it would look like this:

```groovy
import spock.lang.Specification
import spock.lang.Unroll

class AssumptionViolatedTest extends Specification {
    @Unroll
    def "Can ignore test if precondition is not satisfied"() {
        setup: 'check some precondition'
        assumeThat( i == 2 )

        expect:
        i > 0
        where:
        i << [ 1, 2, 3 ]
    }
}
```

With the second proposal:

```groovy
import spock.lang.Specification
import spock.lang.Unroll

class AssumptionViolatedTest extends Specification {
    @Unroll
    def "Can ignore test if precondition is not satisfied"() {
        precondition: 'check some precondition'
        i == 2

        expect:
        i > 0
        where:
        i << [ 1, 2, 3 ]
    }
}
```

Thank you for providing the community with a best-in-class testing framework, and hope that this small addition can help make Spock become even better.

### PR DESCRIPTION
**Pull request by kriegaex**
`@Stepwise` can now be applied to feature methods. The effects are:

  1. the feature is switched to `ExecutionMode.SAME_THREAD`, similarly to how whole specs are switched to the same execution mode when using the annotation on class level.

  2. After the first error or failure occurs in any iteration, all subsequent iterations are going to be skipped.

Fixes #1008.

To do:

- [x] Test automation
- [x] Spock manual update
- [x] Release notes update

_Special thanks to @leonard84 who helped me evolve my separate extension with more complex code structure (because I started implementing it for Spock 1.3 where that was necessary) into something more elegant and minimal. Without him, it would not look quite as beautiful. In the end, I merely had to migrate it from a separate annotation + extension into the existing `StepwiseExtension`, where it is only a small change._

### CODE REVIEW

File: spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java:45]: As discussed in our chat I would prefer using a targeted interceptor (`feature.featureMethod.addInterceptor`), so that we don't add multiple listener to the whole spec.
```groovy
intercept(invocation) {
  try {
    invocation.proceed()
   }
   catch (Throwable e) {
     invocation.feature.skip()
     throw e
   }
}
```

File: spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java
Comment (leonard84) [inline spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java:41]: I would validate that this is indeed a data-driven spec and throw an exception otherwise.

File: docs/release_notes.adoc
Comment (leonard84) [inline docs/release_notes.adoc:7]:
```suggestion
* `@Stepwise` can be applied to data-driven feature methods, having the effect of executing them sequentially (even if concurrent test mode is active) and to skip subsequent iterations if one iteration fails.
```

File: docs/extensions.adoc
Comment (leonard84) [inline docs/extensions.adoc:290]:
```suggestion
Since Spock 2.2, `Stepwise` can be applied to data-driven feature methods, having the effect of executing them sequentially (even if concurrent test mode is active) and to skip subsequent iterations if one iteration fails:
```

File: docs/extensions.adoc
Comment (leonard84) [inline docs/extensions.adoc:303]: Please move this to the docs specs and use include, so that it is actually tested code

File: spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java
Comment (kriegaex) [reply spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java:45]: Sure, I can do that.

But for the protocol: An error listener is only called when an arror occurs, which should be a rare case. A feature method interceptor is always called, i.e. it has to do its "try, catch, proceed" thing for every iteration of the feature it is registered on. Let's say there are 10 iterations, i.e. it is called 10x for every test execution of that feature method. The error listener however, even though it listens for errors in the whole spec, is only called if an error occurs. Let us be generous and say that 1 test fails in 20% of test runs (which would be really flaky), i.e. per test run the error listener is called 0.2x on average. Compare that to 10x. It would be a factor of 50. Even if 1 random test fails during every single test run (what a nightmare!), it is still a factor of 10.

So the reason why I am following your advice is not the efficiency or performance which you seem to have in mind, but the advantage that registering an interceptor on the feature does not need to keep a reference to same feature like the error listener does. It just looks nicer in code, even though it leads to more overhead in many use cases. If you are fine with that, so am I.

File: spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java
Comment (kriegaex) [reply spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java:41]: Why would you do that? The annotation does not make much sense on a non-data-driven feature, but it does not do any harm either. Arguably, there could be a compile error during AST analysis, if you really want to be strict with users, but a runtime error? I would be more lenient there, shall we? But if you insist, saying that Spock should enforce more structure or rules on users, I can do that. I am waiting for your feedback.

File: docs/extensions.adoc
Comment (kriegaex) [reply docs/extensions.adoc:303]: I simply copied the non-tested example code block from a few lines above. Should that one also be converted into something testable and go to the docs specs?

File: spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java
Comment (kriegaex) [reply spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java:41]: BTW, `@Stepwise` on a spec with only a single feature method also does not yield any errors. I think that makes sense, because maybe during TDD the user wants to add more features later, has uncommented some temporarily or whatever. Similarly, we also should not enforce iterations on feature level for `@Stepwise` but simply handle the situation gracefully, which we already do: `skip` is set on the feature, but without any consequences, because there will be no next iteration if we are in the last one already. Having only 1 iteration is but a special case.

File: spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseExtension.groovy
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseExtension.groovy:22]: That prefix grouping doesn't really fit with the rest of the specs, if we want to group something then the usual pattern is two have two separate specs, e.g. `StepwiseExtensionSpecs`/`StepwiseExtensionFeatures`
See `org.spockframework.smoke.extension.IncludeExclude*`

File: spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseExtension.groovy
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseExtension.groovy:112]: You should assert the skip message at least once.

File: spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java
Comment (kriegaex) [reply spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java:41]: The same problem s mentioned in https://github.com/spockframework/spock/pull/1437#issuecomment-1066028196 applies here, I tried. When throwing an exception here, it is **invisible in Maven**, even in debug mode. The test will be silently ignored without even reporting it as ignored. It is as if the test would not exist at all. So probably it is being sorted out by the JUnit platform while aggregating the test plan (or whatever the name for the group of tests to be run is).

The current behaviour is much better IMO. Exceptions thrown by extensions, at least annotation-driven ones (I did not test any global ones in this context yet) are impossible to debug for users like this. Looking at their build logs, they have no way of finding out what is happening. Probably they do not even notice, unless they explicitly look for the spec name in the log or they know the exact number of tests that should be run.

File: docs/extensions.adoc
Comment (kriegaex) [reply docs/extensions.adoc]: Done in 9f2c3a7.

File: spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java
Comment (kriegaex) [reply spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java:49]: Done in 47ceacc, but Maven problem is of course still unsolved, i.e. for now the situation is worse than before where we simply accepted annotations on non-parametrised features, because they had no effect anyway. I am actually feeling bad about this commit and would rather revert it, if you agree. Therefore, not marking this as resolved yet.

File: spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseExtension.groovy
Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseExtension.groovy:112]: Why? The message can change. I am asserting on the number of skipped iterations already. For the spec-level annotation, there is no assertion on the skip message either. I find it interesting that not for the first time, you are requiring something from me which the original code does not have. But anyway, I can comply with that, if you are asking in the spirit of continuous improvement, wishing me to apply the boyscout rule "leave the campground cleaner than you found it". In that case, I will also add a corresponding check in for the spec-based annotation.

File: spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseExtension.groovy
Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseExtension.groovy]: Done in 8d293de.

File: spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseExtension.groovy
Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseExtension.groovy]: Done in 9ef415c.

File: spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java
Comment (kriegaex) [reply spock-core/src/main/java/org/spockframework/runtime/extension/builtin/StepwiseExtension.java:49]: After #1447 has now merged into the main branch, I think I am feeling OK about this change here, so I am resolving the review issue.

**Comment by kriegaex** [comment]
@leonard84, from my perspective, this is done and ready to be merged, unless you find issues in my most recent commits. But better safe than sorry, so take due time to review them.

**Comment by leonard84** [comment]
> I do not know your usual merge practice, so if you want me to rebase on the main branch and/or squash the commits into fewer ones, just let me know. I am not going to rebase or squash anything now, because it might confuse you, reviewing incrementally.

Just commit normally, I normally squash with Github when merging back to master.

File: spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseFeatures.groovy
Comment (leonard84) [inline spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseFeatures.groovy:84]: Will you resolve that?
In `org.spockframework.smoke.CleanupBlocks` a similar pattern is used, although you don't need an inner class for that.

Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseFeatures.groovy:84]: I asked for your feedback concerning this issue in Slack, because I had no better idea. Looking at `CleanupBlocks`, you are using a static member, which is used in two methods. That would not be thread-safe, so you mitigate it using `@Execution(SAME_THREAD)`. I am not sure if that is better. I explicitly wanted my solution to work when running tests concurrently, so I chose a thread-local field. Because I also wanted to avoid the embedded spec to import the complete surrounding spec, I isolated it in an inner static class.

So basically, we are using the same type of solution: We communicate results from the embedded spec to the surrounding class using some kind of static member. That answers my question from Slack if there is a better way, e.g. binding an variable in the embedded spec directly. I guess, this is impossible at present, otherwise you would have done it in `CleanupBlocks`. So my choice would be to not change anything there for now, because my variant should work without `@Execution(SAME_THREAD)`. Would that be OK? If you insist, I can remove the TODO, but actually I like to keep TODOs as reminders for the future and as incentive for co-developers to think about it and come up with a better solution, if the architecture of Spock ever changes in a way enabling such communication more easily, just like with the test results which are communicated to the outside of the spec.

Comment (kriegaex) [reply spock-specs/src/test/groovy/org/spockframework/smoke/extension/StepwiseFeatures.groovy]: For the record after talking privately: We keep that static class with the thread-local (not inherited, though, because according to @leonard84 embedded specs run in the same thread as the feature starting them. The TODO goes away.

**Review by leonard84 — APPROVED** [review]
Thanks for the PR @kriegaex

**Pull request merged**
